File: src/main/java/leetcode/topinterview/LeetCode38.java
package leetcode.topinterview;

/**
 * @author wujingxinit@outlook.com
 * @date 2022/9/22 00:20
 */
public class LeetCode38 {

    public static void main(String[] args) {
        System.out.println(new LeetCode38().countAndSay(7));
    }

    public String countAndSay(int n) {
        StringBuilder sb = new StringBuilder("1");
        for (int i = 1; i < n; i++) {
            StringBuilder temp = new StringBuilder();
            int index = 0;
            Character c = sb.toString().charAt(index);
            int count = 0;
            while (index < sb.length()) {
                if (sb.toString().charAt(index) == c) {
                    count++;
                    if (index == sb.length() - 1) {
                        temp.append(count).append(c);
                    }
                    index++;
                } else {
                    temp.append(count).append(c);
                    c = sb.toString().charAt(index);
                    count = 0;
                }
            }
            sb = temp;
        }
        return sb.toString();
    }
}

File: src/main/go/leetcode/lists/topinterview/LeetCode38.go
package topinterview

import "strconv"

func countAndSay(n int) string {
	if n == 0 {
		return ""
	}
	if n == 1 {
		return "1"
	}
	s := "1"
	for i := 1; i <= n; i++ {
		temp := ""
		ch := s[0]
		count := 0
		index := 0
		for index < len(s) {
			if ch == s[index] {
				index++
				count++
			} else {
				//study
				temp = temp + strconv.Itoa(count) + string(ch)
				ch = s[index]
				count = 1
				index++
			}
		}
		if index == len(s) {
			temp = temp + strconv.Itoa(count) + string(ch)
		}
		s = temp
	}
	return s
}
